### PR TITLE
Add leakage-safe k-sparse probing over activation tensors

### LINKED ISSUE #1728: [Proposal] Add leakage-safe k-sparse probing over activation tensors
### Proposal 

Add a dependency-free, model-free analysis module,
`transformer_lens/tools/analysis/sparse_probing.py`, for fitting binary k-sparse probes to a
supplied activation matrix `X` and label vector `y`.

The first PR contains only:

1. train/test splitting before all learned statistics;
2. train-only mean-difference feature selection;
3. deterministic binary logistic fitting on the selected coordinates;
4. held-out accuracy, precision, recall, and F1;
5. k-sweeps using one fixed split;
6. shuffled-label and random-coordinate controls.

Activation extraction, model wrappers, plotting, notebooks, multiclass probing, and optimal MIP
selection are follow-ups.

---

### Motivation

A central mechanistic-interpretability question is how strongly a labeled feature is concentrated in
individual activation coordinates. Gurnee et al. study this with k-sparse linear probes: select at
most `k` neurons, retrain a classifier on those neurons, and evaluate held-out performance as `k`
changes.

TransformerLens already exposes activations through `run_with_cache`, but has no maintained probing
primitive. Users must repeatedly rebuild splitting, preprocessing, selection, fitting, and control
logic, where leakage is easy to introduce.

The model-free core is also generic machine-learning infrastructure. Whether TransformerLens should
own and maintain a logistic solver, rather than document composition with sklearn, is the primary
repository-fit question for maintainers.

The output supports hypotheses about coordinate concentration. It does **not** establish that the
model uses the decoded feature, that a selected neuron is monosemantic, or that a gradual k-curve is
by itself evidence of superposition.

---

#### Proposed API

Names and optimizer details remain subject to maintainer approval.

```python
from transformer_lens.tools.analysis import (
    fit_sparse_probe,
    sweep_sparse_probe,
)

probe = fit_sparse_probe(
    X,
    y,
    k=4,
    test_fraction=0.3,
    positive_label=1,
    preprocess="none",
    l2_strength=1e-2,
    seed=0,
)

sweep = sweep_sparse_probe(
    X,
    y,
    ks=[1, 2, 4, 8, 16],
    test_fraction=0.3,
    positive_label=1,
    preprocess="none",
    l2_strength=1e-2,
    n_random_subsets=20,
    n_label_shuffles=20,
    seed=0,
)
```

`X` has shape `[example, feature]`; `y` is a Boolean or integer label tensor containing exactly two
values, one equal to `positive_label`. Callers obtain `X` through any appropriate `run_with_cache`
workflow and choose their own position aggregation policy.

#### Design (algorithm)

For `X ∈ R^{n×d}` and binary labels `y`:

1. Validate finite floating `X`, binary Boolean/integer `y`, `1 <= k <= d`,
   `0 < test_fraction < 1`, positive L2 strength, and at least two examples from each class.
2. Create one deterministic stratified split before computing any learned statistic. For class `c`,
   use `n_test_c = clamp(ceil(test_fraction * n_c), 1, n_c - 1)` and report realized counts.
3. On `X_train` only, compute the paper's raw mean-difference score
   `mean(X_train[y=positive]) - mean(X_train[y=negative])`; rank by absolute value with
   deterministic tie-breaking.
4. Select the top `k` coordinates. `preprocess="none"` matches the reference default. Optional
   `"standardize"` uses selected-column train mean and population standard deviation, maps a
   zero standard deviation to scale one, and applies the same transform to test data. Because L2 is
   scale-sensitive, this mode intentionally changes the fitted objective and is recorded in results.
5. Fit logits `z = Xw + b` by minimizing mean class-balanced binary cross-entropy plus
   `l2_strength * ||w||² / 2`; use weights `n / (2 n_c)` and do not regularize `b`.
6. Optimize selected matrices on CPU in float64 with Torch LBFGS and strong-Wolfe line search.
   Require finite output and a configurable final gradient-infinity-norm threshold; otherwise raise.
7. Predict positive at `z >= 0`. Report held-out confusion counts, accuracy, precision, recall, and
   F1, returning zero for precision/F1 when their denominators are zero. F1 is primary.
8. For a sweep, reuse the exact split across every strictly increasing unique `k`. Random-coordinate
   controls sample supports without replacement and fit the same classifier. Label-shuffle controls
   permute training labels, repeat selection/fitting, and evaluate against untouched test labels.

Control repeat counts are nonnegative; zero disables that control. The implementation uses a local
`torch.Generator` and must not mutate global RNG state. It adds no scikit-learn, Gurobi, plotting, or
dataset dependency.

Every k-sweep uses one fixed preprocessing mode and L2 strength. Curve shape is conditional on those
choices; PR1 performs no hyperparameter search and must not present one setting as canonical.

#### Main correctness risk (named)

**Leakage and the probe-fit-vs-model-use conflation.** The single thing most likely to be subtly wrong is **information leaking from test into the probe** — the most common trap is **column standardization or unit-selection computed on the full dataset** (not the train split), which silently inflates held-out accuracy. The close-second is **interpretation**: a high-accuracy probe means the feature is *linearly decodable*, **not** that the model *uses* it — over-claiming here is the field's canonical error.

Mitigations:

- Split first; compute selection and optional normalization on training data only. The function
  cannot detect leakage already introduced into caller-provided `X`, which documentation must state.
- Use deterministic stratification and balanced class weights rather than treating accuracy as safe
  under imbalance.
- Return precision, recall, and F1 next to accuracy.
- Return raw control distributions rather than assigning automatic "significance" labels.
- Document that controls detect some probe-pipeline failures, not causal model use.

#### Validation plan (falsifiable)

1. **Selector oracle.** Independently compute raw train-only class-mean differences and assert exact
   selected indices, signed scores, and deterministic tie-breaking.
2. **Planted sparse feature.** On seeded synthetic data with one known predictive coordinate, `k=1`
   selects it and held-out F1 exceeds a fixed threshold.
3. **Distributed decodability.** On a seeded feature spread across several coordinates, larger `k`
   improves held-out F1 by a fixed margin. The test makes no superposition claim.
4. **Leakage guard.** In standardize mode, hand-computed train-only means/stds and transformed test
   values match metadata; changing only held-out values cannot alter selected coordinates.
5. **Controls.** Label-shuffle and random-coordinate distributions are deterministic for a seed and
   remain below the planted-feature result by fixed margins.
6. **Optimizer.** Objective and gradient match independent formulas; a tiny analytic dataset reaches
   the configured gradient threshold, while forced non-convergence raises.
7. **Validation.** Empty classes, non-binary labels, non-finite values, invalid `k`, infeasible
   splits, duplicate `k` values, and constant selected columns follow explicit tested policies.

#### Scope of the first PR (vertical slice)

Exact paths:

- `transformer_lens/tools/analysis/sparse_probing.py`
- `tests/unit/tools/test_sparse_probing.py`
- exports from `transformer_lens/tools/analysis/__init__.py`
- `docs/source/content/sparse_probing.md`, including composition with `run_with_cache`
- `docs/source/index.md`, adding the guide to the Resources toctree

Public results should include class labels/counts, selected indices/scores, coefficients/intercept,
preprocessing metadata, split indices, held-out metrics, final objective/gradient/iterations, and raw
control distributions. They must not expose a `.plot()` method or assign a representation label.

#### Follow-up work

Deferred: activation extraction, position aggregation, model integration, plotting, notebooks,
multiclass labels, SAE-latent probing, MIP selection, feature batteries, and causal validation.

---

### Alternatives

- **Document sklearn composition instead:** ecosystem-standard and lower maintenance, but leaves
  leakage-sensitive orchestration and result contracts to each user. This may be preferable if
  maintainers do not want TransformerLens to own a classifier optimizer.
- **Depend on the paper repository:** introduces research-code and Gurobi coupling.
- **Ship MIP selection now:** adds a solver dependency and changes the project from a small PR.
- **Bundle model extraction:** forces unresolved prompt/position aggregation policy into the core.

---

### Correctness oracle

The selector oracle is the independently computed train-only raw class-mean difference. The
mean-difference heuristic is **not** required to approximate the exhaustive best subset; no such
guarantee exists. Logistic fitting is checked through analytic planted data, loss/convergence
metadata, and deterministic repetition.

---

### Additional context

- **Paper:** Wes Gurnee et al., ["Finding Neurons in a Haystack: Case Studies with Sparse
  Probing"](https://arxiv.org/abs/2305.01610), TMLR 2023
  ([OpenReview record](https://openreview.net/forum?id=JYs1R9IMJr)).
- **Reference code:** [`wesg52/sparse-probing-paper` pinned at `a610e102`](https://github.com/wesg52/sparse-probing-paper/tree/a610e102c6e25a6ef9cc16c3a2abb736aa90849b):
  [mean-difference selector](https://github.com/wesg52/sparse-probing-paper/blob/a610e102c6e25a6ef9cc16c3a2abb736aa90849b/experiments/probes.py#L168-L195),
  [split and optional scaling](https://github.com/wesg52/sparse-probing-paper/blob/a610e102c6e25a6ef9cc16c3a2abb736aa90849b/experiments/inner_loops.py#L26-L37),
  [heuristic k-sweep](https://github.com/wesg52/sparse-probing-paper/blob/a610e102c6e25a6ef9cc16c3a2abb736aa90849b/experiments/inner_loops.py#L321-L364),
  [experiment defaults](https://github.com/wesg52/sparse-probing-paper/blob/a610e102c6e25a6ef9cc16c3a2abb736aa90849b/config.py#L5-L30), and
  [classification metrics](https://github.com/wesg52/sparse-probing-paper/blob/a610e102c6e25a6ef9cc16c3a2abb736aa90849b/experiments/metrics.py#L14-L35).
- **Paper alignment:** rank coordinates by raw absolute class-mean difference, retrain a balanced
  logistic probe on the selected support, and treat held-out F1 as the primary metric.
- **Verified reference defaults:** the heuristic sweep first creates a random train/test split
  (`test_set_frac=0.3` by default), optionally scales using `X_train` standard deviations, and ranks
  on `X_train`; normalization is disabled by default and, when enabled, divides by clamped standard
  deviation times ten without centering. Its sklearn SAGA fit uses balanced weights, L2, and
  `max_iter=200` (other regularization/tolerance settings use sklearn defaults).
- **Intentional divergences:** stratify, offer explicit centered train-only standardization, use a
  defined Torch loss/L2 convention, add deterministic controls and convergence metadata, and state
  claim boundaries.
- **TransformerLens touchpoints:** current
  [`tools.analysis` package](https://github.com/TransformerLensOrg/TransformerLens/tree/dev/transformer_lens/tools/analysis),
  [`HookedRootModule.run_with_cache`](https://github.com/TransformerLensOrg/TransformerLens/blob/dev/transformer_lens/HookedRootModule.py#L381),
  [`TransformerBridge.run_with_cache`](https://github.com/TransformerLensOrg/TransformerLens/blob/dev/transformer_lens/model_bridge/bridge.py#L2707),
  [`ActivationCache`](https://github.com/TransformerLensOrg/TransformerLens/blob/dev/transformer_lens/ActivationCache.py#L55), and
  [dependency manifest](https://github.com/TransformerLensOrg/TransformerLens/blob/dev/pyproject.toml).
- **Artifact requirements:** none; PR1 tests use only synthetic tensors.

---

### Maintainer decisions requested

1. **Repository fit and scope:** should TransformerLens own a model-free sparse-probe solver at all?
   If yes, should PR1 remain model-free with extraction/plotting deferred? **Recommendation: proceed
   only if maintainers explicitly accept optimizer ownership; then keep PR1 model-free.**
2. **Optimizer:** is the explicitly defined CPU-float64 Torch LBFGS objective acceptable instead of
   adding sklearn? **Recommendation: yes, with final-gradient metadata and failure.**
3. **Controls:** should random-coordinate and shuffled-training-label distributions be first-class
   sweep outputs? **Recommendation: yes, disabled only when repetition count is zero.**
4. **Preprocessing:** should `"none"` match the paper-code default, with optional train-only
   `"standardize"` explicitly documented as changing the L2 objective? **Recommendation: yes.**
5. **Device/dtype:** should score reductions run in at least float32 on the input device, while only
   the selected `[example, k]` matrices move to CPU float64 for deterministic LBFGS?
   **Recommendation: yes; return CPU analysis results and document the transfer explicitly.**

---

### Checklist

- [x] No competing TransformerLens or SAELens probing issue, PR, or implementation found as of
  2026-08-30.
- [x] Paper heuristic and primary metric checked against arXiv:2305.01610.
- [x] Reference defaults and leakage order checked against commit `a610e102`.
- [x] Torch 2.10 CPU-float64 LBFGS prototype converged deterministically across FP16/BF16/FP32/FP64
  inputs; this does not replace tests on the repository's supported Torch/Python matrix.
- [x] PR1 narrowed to model-free fitting, sweeping, controls, tests, exports, and docs.
- [ ] Maintainer decisions recorded before implementation.
- [ ] Follow-up extraction/plotting/model work tracked separately if requested.

---

**Suggested labels:** `enhancement`, `tooling`

---

### PR DESCRIPTION
**Pull request by lorenzozanee**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes 1728

TransformerLens exposed activations but had no maintained k-sparse probing primitive, so leakage-prone splitting and selection were rebuilt per experiment. This adds a leakage-safe, model-free probe that fixes a single stratified split before any learned statistic and reports held-out F1 with deterministic controls.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Screenshots

Please attach before and after screenshots of the change if applicable.

No visual change.

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have not rewritten tests relating to key interfaces which would affect backward compatibility

Reviewers: @jlarson4 @c03d5103 @9a793522

### CODE REVIEW

**Comment by janmenjayap** [comment]
Hi @lorenzozanee, I appreciate the work you have put into this PR. I noticed it while independently working on #1728, and it appears that our implementations overlap substantially. I have paused my branch to avoid creating a duplicate PR

My local draft includes additional work around MPS CPU-first transfers, train-only leakage checks, immutable typed results, deterministic controls, and convergence metadata.

I would be happy either to leave a consolidated review here or to share commits that may be useful to cherry-pick. Please let me know which approach would be most helpful

cc @jlarson4 for guidance on the preferred path forward

**Comment by lorenzozanee** [comment]
Hi @janmenjayap, thank you so much for the thoughtful note — and for pausing your branch. I really appreciate it.

The overlap makes total sense: #1728 asks for something the library clearly needed, and I'm glad we're converging rather than duplicating.

Your additional work — MPS CPU-first transfers, train-only leakage checks, immutable typed results, deterministic controls, and convergence metadata — all sound like real improvements my PR is missing. I'd love to fold them in. If you're willing, please share the commits you think are worth cherry-picking and I'll integrate them (and gladly add you as a co-author). I'm also happy to receive a consolidated review here either way.

Thanks again for being so collaborative — this is the kind of open-source exchange that makes the library better. cc @jlarson4 for any guidance on how you'd like us to proceed.

**Comment by jlarson4** [comment]
Hi @janmenjayap & @lorenzozanee! 

I just posted my maintainer feedback for the original issue's requests, please take a look at those when you have a moment & make sure they're all covered here. I will hold off doing a full review of this PR, as some of my initial comments relate to some of the elements that @janmenjayap mentioned in his comment that this PR is missing. Once you've consolidated that work, ping me and I can give this a full review. In the meantime I am going to focus on getting a review in for @janmenjayap's other open PR, #1723 

Thanks for your hard work on this, keep me posted if there is anything I can do to help from my end

**Comment by lorenzozanee** [comment]
Hi @jlarson4, thank you for the thorough maintainer feedback on #1728 — and for confirming the approach across all five decision requests. Very happy to have the green light on repo fit, the sklearn-free optimizer, the controls design, preprocessing, and device/dtype policy.

I've read through your notes and I'll make sure they're all reflected here. I'm also on board with consolidating the additional work @janmenjayap mentioned. Once that's folded in and the PR is properly updated, I'll ping you for the full review.

Thanks again for holding off and for being so responsive — I'll keep you posted. cc @janmenjayap

**Comment by janmenjayap** [comment]
Hi @lorenzozanee, following our coordination, I have prepared a focused commit directly on top of the current PR head,  0280ac91 .

Commit:  https://github.com/janmenjayap/TransformerLens/commit/cc6470f9425beccafd191b066c0e36f8642a8d0b
Cherry-pick:  `git cherry-pick cc6470f9425beccafd191b066c0e36f8642a8d0b`

The commit:

- Makes indexing and CPU-float64 conversion safe for MPS inputs
- Reports actual LBFGS iterations
- Removes the  `# type: ignore  annotations`
- Aligns convergence thresholds across the API, optimizer, tests, and documentation
- Adds frozen typed result contracts and optimization metadata
- Strengthens held-out leakage and real-MPS regression coverage

Validation passed:

- 8 focused sparse-probing tests
- Real-MPS fit and sweep regression (20 noisy fits and 5 controlled sweeps across seeds)
- 5,316 unit tests
- Full mypy and formatting checks and ocumentation build

Since this commit is based on the current PR head, I suggest cherry-picking it first. Afterward, please rebuild or rebase the branch onto the latest `dev` so the PR contains only the sparse-probing changes and no longer reports merge conflicts. Let me know if any cherry-pick or rebase conflicts arise.

**Comment by lorenzozanee** [comment]
Hi @janmenjayap, thank you — this is extremely helpful.

The commit looks like exactly the kind of focused follow-up we need. I will cherry-pick `cc6470f9425beccafd191b066c0e36f8642a8d0b` onto the current PR head, then rebase/rebuild the branch against the latest `dev` so the PR stays clean and only contains the sparse-probing changes.

I really appreciate the careful validation and the MPS / convergence / typed-result improvements. If any conflicts come up during the cherry-pick or rebase, I will let you know right away.

cc @jlarson4
